extensions/blocks/social-previews/upgrade.js

```diff
@@ -43,10 +59,79 @@ const SocialPreviewsUpgrade = function SocialPreviewsUpgrade() {
 						) }
 					</li>
 				</ul>
-				<Button isPrimary>{ __( 'Upgrade', 'jetpack' ) }</Button>
+
+				<Button
+					href={ href } // Only for server-side rendering, since onClick doesn't work there.
+					isPrimary
+					label={ __( 'Purchase a business plan to access social previews', 'jetpack' ) }
+					onClick={ autosaveAndRedirect }
+					target="_top"
+				>
+					{ __( 'Upgrade', 'jetpack' ) }
+				</Button>
 			</div>
 		</div>
 	);
 };
 
-export default SocialPreviewsUpgrade;
+export default compose( [
+	withSelect( select => {
+		const plan = select( 'wordpress-com/plans' ).getPlan( BUSINESS_PLAN );
+		const postId = select( 'core/editor' ).getCurrentPostId();
+		const postType = select( 'core/editor' ).getCurrentPostType();
+
+		const planPathSlug = plan?.path_slug;
+		const domain = getSiteFragment();
+
+		// The editor for CPTs has an `edit/` route fragment prefixed.
+		const postTypeEditorRoutePrefix = [ 'page', 'post' ].includes( postType ) ? '' : 'edit';
+
+		// Post-checkout: redirect back here.
+		const redirect_to = isSimpleSite()
+			? addQueryArgs(
+					'/' + compact( [ postTypeEditorRoutePrefix, postType, domain, postId ] ).join( '/' ),
+					{
+						plan_upgraded: 1,
+					}
+			  )
+			: addQueryArgs(
+					window.location.protocol + `//${ domain.replace( '::', '/' ) }/wp-admin/post.php`,
+					{
+						action: 'edit',
+						post: postId,
+						plan_upgraded: 1,
+					}
+			  );
+
+		const href =
```

Comment:
This `withSelect` call is becoming quite long and complex.

I wonder whether all the logic for building a redirect link should be extracted to a reusable method. That way you might find it would be easier to extract to the `shared` folder in future if we need to repeat the same logic for other blocks/extensions.

Reply:
In fact, we are duplicating the code now on this PR. It seems to be partially copied from [upgrade-nudge component](https://github.com/Automattic/jetpack/blob/master/extensions/shared/components/upgrade-nudge/index.jsx#L94). I [left a comment](https://github.com/Automattic/jetpack/pull/16797#pullrequestreview-464894697 ) about this some days ago.
In the Paid Blocks PR, we've abstracted this function [here](https://github.com/Automattic/jetpack/blob/9276f88393e16ad77a14e9020970282711b7943d/extensions/shared/plan-utils.js#L41-L76). If we don't refact the code, we are going to triplicate it soon.